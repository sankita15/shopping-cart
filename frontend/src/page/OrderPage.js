import React from 'react';
import { Card, Label, CardText, Col, CardBody, Button, Media, Row, Alert } from 'reactstrap';
import PropTypes from 'prop-types';

export default class OrderPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            cartDetails: {},
            showAlert: false,
            isDisabled: false,
        };
    }

    componentDidMount() {
        const { user } = this.props;

        fetch(`/api/carts/user/${user}`, {
            credentials: 'include',
            headers: {
                'Content-Type': 'application/json',
            },
        }).then(res => (res.ok ? res.json() : Promise.reject(res.status)))
            .then(allCarts => allCarts.filter(pendingCart => pendingCart.status === 'pending')[0])
            .then(cartDetails => ((cartDetails === undefined) ? Promise.reject(404) : this.setState({ cartDetails })))
            .catch(status => console.warn(status));
    }

    isCartEmpty() {
        const { cartDetails } = this.state;
        return Object.keys(cartDetails).length !== 0 && Object.keys(cartDetails.products).length !== 0;
    }

    getProduct = ({ id, productName, productCode, price, imageUrl, starRating }, productQuantity) => {
        const deliveryDate = new Date();

        return (
            <CardBody className="place-order" key={id}>
                <Media src={imageUrl} className="order-image" />
                <CardText className="product-name">
                    {productName}
                </CardText>
                <CardText className="product-code">
                    {productCode}
                </CardText>
                <CardText className="product-price">
                    {`Rs. ${price}`}
                </CardText>
                <CardText className="product-quantity">
                    {`Quantity: ${productQuantity}`}
                </CardText>
                <CardText className="product-rating">
                    {`Star Rating: ${starRating}`}
                </CardText>
                {/* TODO: REFACTOR DATE */}
                <CardText className="product-delivery">
                    {`Estimated Delivery : ${`${deliveryDate.getDate()}/${deliveryDate.getMonth() + 1}/${deliveryDate.getFullYear()}`}`}
                </CardText>
                <div className="horizontal-line" />
            </CardBody>
        );
    };

    getCartProducts() {
        const { cartDetails: { products, productQuantities } } = this.state;

        return Object.keys(products)
            .map(productId => this.getProduct(products[productId], productQuantities[productId]));
    }

    placeOrder() {
        const { cartDetails: { id } } = this.state;

        fetch(`/api/carts/order/${id}`, {
            credentials: 'include',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
        }).then(res => (res.ok ? res.json() : Promise.reject(res.status)))
            .then(() => {
                this.setState({ showAlert: true, isDisabled: true });
            }).catch(status => console.warn(status));
    }

    productOrderedAlert() {
        const { showAlert } = this.state;
        return (
            <div>
                <Alert color="success" className="alert-box" isOpen={showAlert} toggle={() => this.onDismiss()}>
                Yayy!!! Your Order Got Placed.
                Click on cancel to continue shopping
                </Alert>
            </div>
        );
    }

    render() {
        const { user } = this.props;
        const { showAlert, isDisabled } = this.state;
        const delivery = 40;

        if (!this.isCartEmpty()) {
            return <Label>Your Cart is empty.Please add item to your cart</Label>;
        }

        return (
            <div>
                <Row>{ showAlert && this.productOrderedAlert()}</Row>
                <Label className="welcome-order">{`Hello ${user.toUpperCase()}, Review Your Order`}</Label>
                <Row>
                    <Col sm={8}>
                        <Card className="order-card">
                            {
                                this.getCartProducts()
                            }
                        </Card>
                    </Col>
                    <Col sm={3}>
                        <Card className="order-value-card">
                            <CardBody className="order-price">
                                <CardText className="order-summary">Order Summary</CardText>
                                <CardText>{`Items: ${this.getTotalPrice()}`}</CardText>
                                <CardText>
                                    {' '}
                                    {`Delivery: Rs ${delivery}`}
                                    {' '}
                                </CardText>
                                <CardText>
                                    {' '}
                                    {`Cart Total: Rs ${this.getTotalPrice() + delivery}`}
                                    {' '}
                                </CardText>
                                <Button className="place-order-button" disabled={isDisabled} onClick={() => this.placeOrder()}>Place Your Order</Button>
                            </CardBody>
                        </Card>
                    </Col>
                </Row>
            </div>
        );
    }

    onDismiss() {
        this.setState({
            showAlert: false,
        });
        this.redirectToProductPage();
    }

    getTotalPrice() {
        const { cartDetails: { totalPrice } } = this.state;
        return totalPrice;
    }

    redirectToProductPage = () => window.location.assign('/products');
}

OrderPage.propTypes = {
    user: PropTypes.string.isRequired,
};
