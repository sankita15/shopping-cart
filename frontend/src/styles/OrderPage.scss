.order-card{
  margin-left: 30px;
  margin-top: 40px;
}
.order-value-card{
  margin-top: 40px;
}
.order-summary{
  font-weight: bold;
}
.place-order-button{
  background-color: #007bff;
  border-color: #007bff;
  width: 100%;
}
.welcome-order{
  margin-top: 58px;
  margin-left: 30px;
  font-weight: bold;
}
.horizontal-line{
  background-color: lightgrey;
  height: 1px;
}
.order-image{
  height : 77px;
}
.product-name{
  font-weight: bold;
  margin-left: 127px;
  margin-top: -78px;
}
.product-code{
  margin-left: 127px;
  margin-top: -20px;
}
.product-price{
  margin-left: 127px;
  margin-top: -20px;
}
.product-quantity{
  margin-left: 127px;
  margin-top: -19px;
}
.product-rating{
  margin-left: 127px;
  margin-top: -19px;
}
.product-delivery{
  margin-left: 127px;
  margin-top: -19px;
  font-weight: bold;
}
.alert-box{
  width: 1453px;
}
