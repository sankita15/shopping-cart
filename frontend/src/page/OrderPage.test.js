import React from 'react';
import { shallow } from 'enzyme';
import { Col, Card, Label, Button, CardBody, CardText, Alert } from 'reactstrap';
import OrderPage from './OrderPage';


describe('Order Page', () => {
    const defaultProps = {
        user: 'alice',
    };

    const CART = [{
        id: '5c6bf8a306057eeb12dfb7c4',
        status: 'pending',
        products: {
            '59d89875479ded2a718bac13': {
                'id': '59d89875479ded2a718bac13',
                'productName': 'colgatewhite',
                'productCode': 'GE-1206',
                'releaseDate': 1483539355570,
                'lastModified': 1507367029712,
                'description': 'dentifrice',
                'price': 300,
                'starRating': 4.5,
                'imageUrl': 'http://colgate.com/2/',
                'stock': 1000,
            },
            '59d89875479ded2a718bac14': {
                'id': '59d89875479ded2a718bac14',
                'productName': 'gilettemousse',
                'productCode': 'GE-1206',
                'releaseDate': 1486224955570,
                'lastModified': 1507367029713,
                'description': 'moussearaser',
                'price': 200,
                'starRating': 4.2,
                'imageUrl': 'http://gilette.com/2/',
                'stock': 1000,
            },
        },
        productQuantities: {
            '59d89875479ded2a718bac13': 2,
            '59d89875479ded2a718bac14': 2,
        },
        lastModified: 1551162037567,
        orderDate: 1550482348428,
        totalPrice: 1000,
    }];

    const ORDERED_CART = [{
        id: '5c6bf8a306057eeb12dfb7c4',
        status: 'ordered',
        products: {
            '59d89875479ded2a718bac13': {
                'id': '59d89875479ded2a718bac13',
                'productName': 'colgatewhite',
                'productCode': 'GE-1206',
                'releaseDate': 1483539355570,
                'lastModified': 1507367029712,
                'description': 'dentifrice',
                'price': 300,
                'starRating': 4.5,
                'imageUrl': 'http://colgate.com/2/',
                'stock': 1000,
            },
            '59d89875479ded2a718bac14': {
                'id': '59d89875479ded2a718bac14',
                'productName': 'gilettemousse',
                'productCode': 'GE-1206',
                'releaseDate': 1486224955570,
                'lastModified': 1507367029713,
                'description': 'moussearaser',
                'price': 200,
                'starRating': 4.2,
                'imageUrl': 'http://gilette.com/2/',
                'stock': 1000,
            },
        },
        productQuantities: {
            '59d89875479ded2a718bac13': 2,
            '59d89875479ded2a718bac14': 2,
        },
        lastModified: 1551162037567,
        orderDate: 1550482348428,
        totalPrice: 1000,
    }];

    const createWrapper = () => shallow(<OrderPage {...defaultProps} />);

    window.location.assign = jest.fn();

    let component;

    beforeEach(async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(CART), ok: true });

        component = createWrapper();

        // eslint-disable-next-line no-undef
        await flushPromises();
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should render Order page', () => {
        const welcomeMessage = component.find(Label);
        const productsOrdered = component.find(Col).at(0).find(Card);
        const orderDetails = component.find(Col).at(1).find(Card);
        const buyNowButton = component.find(Col).at(1).find(Card).find(Button);

        expect(welcomeMessage.length).toEqual(1);
        expect(welcomeMessage.props().children).toEqual(`Hello ${defaultProps.user.toUpperCase()}, Review Your Order`);
        expect(productsOrdered.length).toEqual(1);
        expect(productsOrdered.find(CardBody).find(CardText).at(0).props().children).toEqual('colgatewhite');
        expect(orderDetails.length).toEqual(1);
        expect(orderDetails.find(CardBody).find(CardText).at(0).props().children).toEqual('Order Summary');

        expect(buyNowButton.length).toEqual(1);
    });

    it('should call fetch for getting all carts for a particular user', () => {
        expect(fetch).toHaveBeenCalledWith(`/api/carts/user/${defaultProps.user}`, {
            credentials: 'include',
            headers: {
                'Content-Type': 'application/json',
            },
        });
    });

    it('should fetch /order api on click of place your order button', async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(ORDERED_CART), ok: true });

        const buyNowButton = component.find(Col).at(1).find(Card).find(Button);

        expect(component.find(Alert).length).toEqual(0);

        buyNowButton.simulate('click');

        component.update();

        // eslint-disable-next-line no-undef
        await flushPromises();

        expect(fetch).toHaveBeenCalledWith(`/api/carts/order/${CART[0].id}`, {
            credentials: 'include',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
        });

        const alertBox = component.find(Alert);

        expect(alertBox.length).toEqual(1);

        alertBox.props().toggle();

        expect(window.location.assign).toHaveBeenCalledWith('/products');
    });

    it('should disable button after order gets placed', async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(ORDERED_CART), ok: true });

        const buyNowButton = component.find(Col).at(1).find(Card).find(Button);

        expect(buyNowButton.prop('disabled')).toBeFalsy();

        buyNowButton.simulate('click');

        component.update();

        // eslint-disable-next-line no-undef
        await flushPromises();

        const buyNowButton1 = component.find(Col).at(1).find(Card).find(Button);

        expect(buyNowButton1.prop('disabled')).toBeTruthy();
    });
});
