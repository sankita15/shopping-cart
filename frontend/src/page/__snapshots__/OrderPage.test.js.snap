// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Page should match snapshot 1`] = `
<div>
  <Row
    tag="div"
  />
  <Label
    className="welcome-order"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    Hello ALICE, Review Your Order
  </Label>
  <Row
    tag="div"
  >
    <Col
      sm={8}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Card
        className="order-card"
        tag="div"
      >
        <CardBody
          className="place-order"
          key="59d89875479ded2a718bac13"
          tag="div"
        >
          <Media
            className="order-image"
            src="http://colgate.com/2/"
          />
          <CardText
            className="product-name"
            tag="p"
          >
            colgatewhite
          </CardText>
          <CardText
            className="product-code"
            tag="p"
          >
            GE-1206
          </CardText>
          <CardText
            className="product-price"
            tag="p"
          >
            Rs. 300
          </CardText>
          <CardText
            className="product-quantity"
            tag="p"
          >
            Quantity: 2
          </CardText>
          <CardText
            className="product-rating"
            tag="p"
          >
            Star Rating: 4.5
          </CardText>
          <CardText
            className="product-delivery"
            tag="p"
          >
            Estimated Delivery : 26/2/2019
          </CardText>
          <div
            className="horizontal-line"
          />
        </CardBody>
        <CardBody
          className="place-order"
          key="59d89875479ded2a718bac14"
          tag="div"
        >
          <Media
            className="order-image"
            src="http://gilette.com/2/"
          />
          <CardText
            className="product-name"
            tag="p"
          >
            gilettemousse
          </CardText>
          <CardText
            className="product-code"
            tag="p"
          >
            GE-1206
          </CardText>
          <CardText
            className="product-price"
            tag="p"
          >
            Rs. 200
          </CardText>
          <CardText
            className="product-quantity"
            tag="p"
          >
            Quantity: 2
          </CardText>
          <CardText
            className="product-rating"
            tag="p"
          >
            Star Rating: 4.2
          </CardText>
          <CardText
            className="product-delivery"
            tag="p"
          >
            Estimated Delivery : 26/2/2019
          </CardText>
          <div
            className="horizontal-line"
          />
        </CardBody>
      </Card>
    </Col>
    <Col
      sm={3}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Card
        className="order-value-card"
        tag="div"
      >
        <CardBody
          className="order-price"
          tag="div"
        >
          <CardText
            className="order-summary"
            tag="p"
          >
            Order Summary
          </CardText>
          <CardText
            tag="p"
          >
            Items: 1000
          </CardText>
          <CardText
            tag="p"
          >
             
            Delivery: Rs 40
             
          </CardText>
          <CardText
            tag="p"
          >
             
            Cart Total: Rs 1040
             
          </CardText>
          <Button
            className="place-order-button"
            color="secondary"
            disabled={false}
            onClick={[Function]}
            tag="button"
          >
            Place Your Order
          </Button>
        </CardBody>
      </Card>
    </Col>
  </Row>
</div>
`;
